[Teoh Jun Jie] iP

junjieteoh opened the pull request:
# Duke

> “Your mind is for having ideas, not holding them.” – David Allen ([source]( https://dansilvestre.com/productivity-quotes))

Duke frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ _SUPER_  FAST to use 

All you need to do is,
1. download it from [here](https://github.com/junjieteoh/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😜 

And it is **FREE** 👍 !

Features: 
- [x] Managing tasks
- [x] Managing deadlines 
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

junjieteoh committed:
Added Greet, Echo, Exit features

junjieteoh committed:
Add the ability to store user text, created Task class

junjieteoh committed:
Updated data structure of arrayOfList to a Task array

junjieteoh committed:
Added the mark as done feature

junjieteoh committed:
Added support for tracking three types of tasks (todo, event, deadline)

junjieteoh committed:
Updated test cases

junjieteoh committed:
Added users input error handling

junjieteoh committed:
Added delete features

junjieteoh committed:
Updated testcases to include features up to Level 6

junjieteoh committed:
Add file writing feature to data/duke.txt

junjieteoh committed:
Add LocalDate objects

junjieteoh committed:
Fix typo

junjieteoh committed:
Add dateObject

junjieteoh committed:
Merge branch 'branch-Level-7'

junjieteoh committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/Task.java

junjieteoh committed:
Add Ui class

junjieteoh committed:
Add Storage class

junjieteoh committed:
Update test cases

junjieteoh committed:
Divide classes into pacakges

junjieteoh committed:
Add task class

junjieteoh committed:
Add JUnit tests

junjieteoh committed:
Add JavaDoc comments

junjieteoh committed:
Revert "Add JavaDoc comments"

This reverts commit e9d779d77ff510d4be65486295d2747a5aab2c24.

junjieteoh committed:
Tweak the code to comply with a coding standard

junjieteoh committed:
Merge branch 'branch-A-CodingStandard'

junjieteoh committed:
Add JavaDoc

junjieteoh committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/parser/DateTimeParser.java

junjieteoh committed:
Add find feature

junjieteoh committed:
Update javadoc for find features

junjieteoh committed:
Update testcase

junjieteoh committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

junjieteoh committed:
Update gradle file.

cpwill01 commented on src/main/java/duke/parser/DateTimeParser.java:
```diff
@@ -0,0 +1,35 @@
+
+
+    /**
+     * Return a LocalDate object based on the input String
```
According to the part of the coding standard on [Javadoc headers](https://se-education.org/guides/conventions/java/intermediate.html#comments), I think this should start with "Returns" rather than "Return". I found similar issues with the header comments in other classes as well.

cpwill01 commented on src/main/java/duke/parser/DateTimeParser.java:
```diff
@@ -0,0 +1,35 @@
+     * @return a LocalDate object from the parsed String.
+     * @throws DateTimeParseException
+     */
+
```
Should this empty line be deleted?

cpwill01 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,214 @@
+     *
+     * @return False if the command is "bye", True otherwise
+     */
+    public boolean processInput() {
```
Should this return a boolean? Based on my understanding, I think methods that return a boolean should also follow the `isXYZ()` naming convention (eg. `isExit()`, `isDone()`).

cpwill01 replied:
Perhaps you could consider separating this into 2 functions where 1 solely handles displaying the correct response while another handles checking whether the application should exit?

cpwill01 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,214 @@
+
+        updateFile();
+
+        String output = "   __________________________________________________\n"
```
Should this line and line 204 below use `Ui.createLine()` instead?

cpwill01 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,214 @@
+     * @return False if the command is "bye", True otherwise
+     */
+    public boolean processInput() {
+        if(userInput.equals("bye")) {
```
There should be a space between `if` and the open bracket. I noticed the same issue on line 81 and in some of the other files. Perhaps you could consider using [checkStyle](https://nus-cs2103-ay2122s2.github.io/website/schedule/week4/project.html#4-add-increments-as-branches-a-gradle-a-checkstyle-level-10-a-varargs)? It really helps me a lot in fixing my code to meet the coding standards 😅

junjieteoh committed:
Add a GUI to Duke.

junjieteoh committed:
Update build.gradle

cpwill01 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,214 @@
+                String newTaskMessage = Parser.createTask(userInput);
+                System.out.println(newTaskMessage);
+            } catch (InvalidTypeException e) {
+                String errorMsg = Ui.createLine()
+                        + "       Opps, the command \"" + userInput + "\" is not supported :(\n"
+                        + Ui.createLine();
+                System.out.println(errorMsg);
```
Would there be a way to use the error messages stored inside the exceptions instead of creating a new error message here? Otherwise, is there a need for the exceptions to have a message in their constructors?

cpwill01 replied:
Would there be a way to use the error messages stored inside the exceptions instead of creating a new error message here? Otherwise, is there a need for the exceptions to have a message in their constructors?

junjieteoh committed:
Merge branch 'branch-Level-10'

cpwill01 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,146 @@
+    protected String name;
+    protected boolean isDone;
+    protected int id;
+    private static ArrayList<Task> listOfTask = new ArrayList<>(100);
```
Should this be called `tasks` instead of `listOfTask`? [Plural form](https://se-education.org/guides/conventions/java/intermediate.html#naming) should be used for names representing collections.

cpwill01 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,146 @@
+     *
+     * @return a String with the current task list
+     */
+    public static String printArray() {
```
Should this and the overloaded method below be called `printTaskList()` instead of `printArray()`? I got a little confused when I saw `Task.printArray()` being used in the other classes because I wasn't sure what array it's referring to.

cpwill01 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,146 @@
+    }
+
+    /**
+     * Retrieve a formatted String of the current task list.
```
Should this description be changed? I understood the method as retrieving the string of only the part of the task list specified by the input array rather than the full task list.

cpwill01 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,146 @@
+     * @param array an integer array containing task number to be printed
+     * @return a String with the current task list
+     */
+    public static String printArray(Integer[] array) {
```
Could the parameter name be more descriptive rather than `array`?

cpwill01 reviewed (commented):
Overall, I found your code mostly easy to follow. However, I did spot some coding standard violations. I've pointed out some of them, but I may have missed some out so I really recommend setting up checkStyle as it helped me spot all the small errors in my own code. There were also some code quality issues I pointed out, but most of them should be quick to fix. All the best for the rest of the iP and tP! 😄

HanJiyao commented on src/main/java/duke/parser/DateTimeParser.java:
```diff
@@ -0,0 +1,35 @@
+import java.time.format.DateTimeParseException;
+import java.time.format.FormatStyle;
+import java.time.LocalDate;
+
+public class DateTimeParser {
+
```
maybe can add missing class JavaDoc here?

HanJiyao commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,57 @@
+        }
+    }
+
+    public static String getCurrentFilePath() {
+        return filePath;
+    }
```
maybe this method is never used?

HanJiyao commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,146 @@
+    protected boolean isDone;
+    protected int id;
+    private static ArrayList<Task> listOfTask = new ArrayList<>(100);
+    private static int counter = 0;
+
```
Since you define list of task as ArrayList, is it better to check array size directly?

HanJiyao commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,41 @@
+     * Print divider line.
+     * @return divider line
+     */
+    public static String createLine() {
+        return  "   __________________________________________________\n";
+    }
```
Should it can represent this line as constant DIVIDER to be more straight forward as it is only one String?

HanJiyao commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,22 @@
+    @Override
+    public String toString() {
+        String status = this.isDone ? "X" : " ";
+        return "[" + this.type + "][" + status + "]" + this.name + " (by: "
+                + DateTimeParser.formatDate(this.time) + ")";
```
according to code standard the breaks should occurs outside the parenthesized expression
longName1 = longName2 * (longName3 + longName4 - longName5)
        + 4 * longname6

HanJiyao reviewed (commented):
Overall, I found your code easy to read for the most part. I noted a few minor coding standard violations too. Some of the classes are getting quite long. Consider splitting into
smaller classes like different command if that makes sense.

junjieteoh committed:
Insert assertions

Currently there are no assertions in the Parser class.

The lack of assertions causes certain assumptions that are needed to be hold true to be missed.

Some assertions to do input validation checks were added for the "mark", "unmark" and "delete" commands.

junjieteoh committed:
Refactor code

There are several violations for code convention and quality.

Refactoring is needed to make the code more readible.

A new class Command has been added.

junjieteoh committed:
Upgrade search feature for better keyword searching.

junjieteoh committed:
Implement better gui

junjieteoh committed:
Update User Guide

junjieteoh committed:
Set theme jekyll-theme-slate

junjieteoh committed:
Update user guide

junjieteoh committed:
Merge branch 'master' of https://github.com/junjieteoh/ip

junjieteoh committed:
Update README.md

junjieteoh committed:
Update README.md
